user-management/src/main/java/de/tum/in/www1/artemis/usermanagement/service/connectors/jira/JiraAuthenticationProvider.java

```diff
@@ -0,0 +1,320 @@
+            }
+        }
+
+        if (authenticationResponse != null && authenticationResponse.getBody() != null) {
+            final var jiraUserDTO = authenticationResponse.getBody();
+            username = jiraUserDTO.getName();
+            User user;
+            final var optionalUser = userRepository.findOneWithGroupsAndAuthoritiesByLogin(username);
+            if (optionalUser.isPresent()) {
+                user = optionalUser.get();
+            }
+            else {
+                // the user does not exist yet, we have to create it in the Artemis database
+                // Note: we use an empty password, so that we don't store the credentials of Jira users in the Artemis DB (Spring enforces a non null password)
+                user = userCreationService.createInternalUser(username, "", null, jiraUserDTO.getDisplayName(), "", jiraUserDTO.getEmailAddress(), null, null, "en");
+                // load additional details if the ldap service is available and the user follows the allowed username pattern (if specified)
+                ldapUserService.ifPresent(service -> service.loadUserDetailsFromLdap(user));
+            }
+            final var groups = jiraUserDTO.getGroups().getItems().stream().map(JiraUserGroupDTO::getName).collect(Collectors.toSet());
+            user.setGroups(groups);
+            user.setAuthorities(authorityService.buildAuthorities(user));
+
+            if (!user.getActivated()) {
+                user.setActivated(true);
+                user.setActivationKey(null);
+            }
+            return userCreationService.saveUser(user);
+        }
+        else {
+            throw new InternalAuthenticationServiceException("JIRA Authentication failed for user " + username);
+        }
+    }
+
+    /**
+     * Adds a JIRA user to a JIRA group. Ignores "user is already a member of" errors.
+     *
+     * @param user     The user
+     * @param group    The JIRA group name
+     * @throws ArtemisAuthenticationException if JIRA returns an error
+     */
+    @Override
+    public void addUserToGroup(User user, String group) throws ArtemisAuthenticationException {
+        // then we also make sure to add it into JIRA so that the synchronization during the next login does not remove the group again
+        log.info("Add user {} to group {} in JIRA", user.getLogin(), group);
+        if (!isGroupAvailable(group)) {
+            throw new IllegalArgumentException("Jira does not have a group: " + group);
+        }
+
+        final var body = new JiraUserDTO(user.getLogin());
+        HttpEntity<?> entity = new HttpEntity<>(body);
+        try {
+            restTemplate.exchange(jiraUrl + "/rest/api/2/group/user?groupname=" + group, HttpMethod.POST, entity, Void.class);
+        }
+        catch (HttpClientErrorException e) {
+            if (e.getStatusCode().equals(HttpStatus.BAD_REQUEST) && e.getResponseBodyAsString().contains("user is already a member of")) {
```

Comment:
Is there any other way to detect that the error is `user is already a member of`?
I think this logic may fail if the error message will change in the future. Does this type of exception have any other information like `error_type` of `error_id` or something similar?

Reply:
Unfortunately, I think there is no other way because the same error is sent when the user is not available and the only difference is in the message. I didn't find additional information like error type or error id.